src/halfpipe/tui/feature_widgets/base.py

```diff
@@ -0,0 +1,474 @@
+
+        occupied_feature_names = [self.feature_items[item].name for item in self.feature_items]
+        await self.app.push_screen(
+            FeatureNameInput(occupied_feature_names),
+            action_rename_feature,
+        )
+
+    @on(Button.Pressed, "#sidebar .duplicate_button")
+    async def duplicate(self) -> None:
+        await self.run_action("duplicate_feature")
+
+    @on(Button.Pressed, "#sidebar .sort_button")
+    async def sort(self) -> None:
+        await self.run_action("sort_features")
+
+    def action_add_feature(self) -> None:
+        # Try here first the event files
+        # setting_filter_step_instance = SettingFilterStep()
+        # setting_filter_step_instance.run()
+        #  events_type_instance = EventsTypeStep()
+        #  events_type_instance.run()
+
+        """Pops out the feature type selection windows and then uses add_new_feature function to mount a new feature
+        widget."""
+        occupied_feature_names = [self.feature_items[item].name for item in self.feature_items]
+        self.app.push_screen(
+            FeatureSelectionScreen(occupied_feature_names),
+            self.add_new_feature,
+        )
+
+    async def add_new_feature(self, new_feature_item: tuple | bool) -> None:
+        """Principle of adding a new feature lies in mounting a new widget while creating a new entry in the dictionary
+        to keep track of the selections which are later dumped into the Context object.
+        If this is a load or a duplication, then new entry is not created but read from the dictionary.
+        The dictionary entry was created elsewhere.
+        """
+        # print("new_feature_itemnew_feature_itemnew_feature_itemnew_feature_item", new_feature_item)
+        if isinstance(new_feature_item, tuple) or isinstance(new_feature_item, list):
+            feature_type, feature_name = new_feature_item
+            new_id = "feature_item_" + str(self._id_counter)
+            self.feature_items[new_id] = FeatureItem(feature_type, feature_name)
+            # the pseudo class here is to set a particular text color in the left panel
+            new_list_item = ListItem(
+                Label(feature_name, id=new_id, classes="labels " + feature_type),
+                id=new_id,
+                classes="items",
+            )
+            # this dictionary will contain all made choices
+            if feature_name not in ctx.cache:
+                if feature_type != "preprocessed_image":
+                    ctx.cache[feature_name]["features"]["name"] = feature_name
+                    ctx.cache[feature_name]["features"]["setting"] = feature_name + "Setting"
+                    ctx.cache[feature_name]["settings"]["name"] = feature_name + "Setting"
+                else:
+                    ctx.cache[feature_name]["settings"]["name"] = feature_name + "UnfilteredSetting"
+                    ctx.cache[feature_name]["settings"]["output_image"] = True
```

Comment:
I would swap the two blocks for ease of reading. Also, why is the setting for `preprocessed_image` called unfiltered? I think we only need to add an unfiltered setting for FALFF.
```suggestion
                if feature_type == "preprocessed_image":
                    ctx.cache[feature_name]["settings"]["name"] = feature_name + "UnfilteredSetting"
                    ctx.cache[feature_name]["settings"]["output_image"] = True
                else:
                    ctx.cache[feature_name]["features"]["name"] = feature_name
                    ctx.cache[feature_name]["features"]["setting"] = feature_name + "Setting"
                    ctx.cache[feature_name]["settings"]["name"] = feature_name + "Setting"
```